Add Locks functionality to UI.

hatched opened the pull request:
This PR adds the UI and associated backend functionality for creating, displaying and removing locks from the UI.

[e companion](https://github.com/gravitational/teleport.e/pull/1029)

<img width="1556" alt="Screenshot 2023-03-20 at 9 53 25 AM" src="https://user-images.githubusercontent.com/532033/226396336-059c2fa2-ea83-45d2-98b3-b4db2f7d14ee.png">
<img width="440" alt="Screenshot 2023-03-20 at 10 58 11 AM" src="https://user-images.githubusercontent.com/532033/226413447-366a3ecd-6efd-45ab-945a-c1c2d3698050.png">
<img width="1555" alt="Screenshot 2023-03-20 at 9 54 09 AM" src="https://user-images.githubusercontent.com/532033/226396351-09bda26a-1d26-4a7b-a727-b8a3e9aad87d.png">
<img width="510" alt="Screenshot 2023-03-20 at 1 10 09 PM" src="https://user-images.githubusercontent.com/532033/226441982-ac5df574-7d7a-4f2c-b682-9d90f0a645cc.png">
<img width="1555" alt="Screenshot 2023-03-20 at 9 53 54 AM" src="https://user-images.githubusercontent.com/532033/226396346-d9d0aef0-90bc-460e-8050-8d38f3136303.png">

hatched committed:
Adds a new section to the UI for managing locks.
New endpoints were added to the API server for fetching, creating and deleting locks.
The 'editor' role now has the ability to create, edit, and remove locks by default.
Created new SlidePanel component to easily add a panel that slides in from the right of the screen.

hatched commented on lib/services/presets.go:
```diff
@@ -204,6 +207,7 @@ func defaultAllowRules() map[string][]types.Rule {
 			types.NewRule(types.KindSAMLIdPServiceProvider, RW()),
 			types.NewRule(types.KindOktaImportRule, RW()),
 			types.NewRule(types.KindOktaAssignment, RW()),
+			types.NewRule(types.KindLock, RW()),
```
This adds the ability for users with the `editor` default role to be able to work with locks. Other users will need the lock resource rules added.

hatched commented on web/packages/build/index.ejs:
```diff
@@ -11,7 +11,7 @@
     <script src="/web/config.js"></script>
   </head>
-  <body>
+  <body style="overflow: hidden">
```
These overflows were added to support the new sliding panel component.

hatched commented on web/packages/design/src/SlidePanel/SlidePanel.tsx:
```diff
@@ -0,0 +1,83 @@
+/*
```
This component can replace the one used in the access request component once merged.

hatched commented on web/packages/teleport/src/Locks/useGetTargetData.tsx:
```diff
@@ -0,0 +1,149 @@
+
+import { useEffect, useState, useMemo } from 'react';
+
+import useTeleport from 'teleport/useTeleport';
```
mocking `useTeleport` is problematic due to its internal structure, we probably want to rearchitect this sometime in the future.

avatus commented on web/packages/teleport/src/Locks/CreateLock.tsx:
```diff
@@ -0,0 +1,171 @@
+            // It takes longer for the cache to be updated when adding locks so
+            // this waits 1s before redirecting to fetch the list again.
+            history.push(cfg.getLocksRoute(clusterId));
+          }, 1000);
```
I don't think we block this way anywhere else in the UI do we?

hatched replied:
We don't, but I also wasn't able to find another instance where an action that was taken in the UI would result in a delay while we waited for cache propagation. 

I'm not too happy with the necessity of this fix... 

The other option would be to inject the new item (or remove the deleted one) from the UI. This however had an issue where the timestamps would differ when the UI updated which felt like it would be a little odd from the user perspective.

zmb3 replied:
Will the UI at least stay in some sort of loading state during this time?

hatched replied:
These are all good points, I'll add a loading state while this timeout is in effect.

hatched replied:
Loading states have been added.

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -2355,6 +2360,132 @@ func createIdentityContext(login string, sessionCtx *SessionContext) (srv.Identi
+	sessionCtx *SessionContext,
+	site reversetunnel.RemoteSite) (interface{}, error) {
+	ctx := r.Context()
+	clt, err := sessionCtx.GetUserClient(r.Context(), site)
```
```suggestion
	clt, err := sessionCtx.GetUserClient(ctx, site)
```

Use `ctx` or `r.Context()` consistently. Switching back and forth between the two is confusing.

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -2355,6 +2360,132 @@ func createIdentityContext(login string, sessionCtx *SessionContext) (srv.Identi
+	// sort or filter results.
+	lockList := make([]UILock, 0, len(locks))
+	for i := 0; i < len(locks); i++ {
+		expires := ""
+		createdAt := ""
```
```suggestion
		var expires, createdAt string
```

Up to you, but I find this a little cleaner.

hatched replied:
I agree, updated.

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -2355,6 +2360,132 @@ func createIdentityContext(login string, sessionCtx *SessionContext) (srv.Identi
+	// Table component doesn't support nested complex objects. And fails to properly
+	// sort or filter results.
+	lockList := make([]UILock, 0, len(locks))
+	for i := 0; i < len(locks); i++ {
```
Why not use a `range` loop?

hatched replied:
Habit I suppose 😁 updated!

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -2355,6 +2360,132 @@ func createIdentityContext(login string, sessionCtx *SessionContext) (srv.Identi
+	p httprouter.Params,
+	sessionCtx *SessionContext,
+	site reversetunnel.RemoteSite) (interface{}, error) {
+	var req *createLockReq
+
+	if err := httplib.ReadJSON(r, &req); err != nil {
```
```suggestion
	var req *createLockReq
	if err := httplib.ReadJSON(r, &req); err != nil {
```

I'd remove this empty line, the req is closely related to the following line which populates it.

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -2355,6 +2360,132 @@ func createIdentityContext(login string, sessionCtx *SessionContext) (srv.Identi
+	}
+
+	ctx := r.Context()
+	clt, err := sessionCtx.GetUserClient(r.Context(), site)
```
```suggestion
	clt, err := sessionCtx.GetUserClient(ctx, site)
```

similarly

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -2355,6 +2360,132 @@ func createIdentityContext(login string, sessionCtx *SessionContext) (srv.Identi
+		}
+	}
+
+	var ttl *time.Time
```
1. Why a pointer? 
2. `ttl` is a poor name for a `time.Time` because "time to live" is typically a duration, not a specific time.

hatched replied:
The `Expires` key in `LockSpecV2` is a pointer. We could change it to match the recently added `CreatedAt` field but I felt that was out of scope for this PR.

This naming is a bit confusing I'll change it to `expires`.

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -2355,6 +2360,132 @@ func createIdentityContext(login string, sessionCtx *SessionContext) (srv.Identi
+	sessionCtx *SessionContext,
+	site reversetunnel.RemoteSite) (interface{}, error) {
+	ctx := r.Context()
+	clt, err := sessionCtx.GetUserClient(r.Context(), site)
```
```suggestion
	clt, err := sessionCtx.GetUserClient(ctx, site)
```

zmb3 commented on web/packages/teleport/index.html:
```diff
@@ -10,7 +10,7 @@
     <title></title>
   </head>
-  <body>
+  <body style="overflow: hidden">
```
This is a pretty global setting. Are we confident we didn't break the UI in any other places?

zmb3 commented on web/packages/teleport/src/Locks/CreateLock.test.tsx:
```diff
@@ -0,0 +1,138 @@
+  useLocks: () => ({ createLock: mockCreateLock }),
+}));
+
+describe('component: CreateLock', () => {
```
+1 for real behavioral tests!

zmb3 commented on web/packages/teleport/src/config.ts:
```diff
@@ -193,6 +195,9 @@ const cfg = {
     mfaDevicePath: '/v1/webapi/mfa/token/:tokenId/devices/:deviceName',
 
+    locksPath: '/v1/webapi/sites/:clusterId/locks',
+    locksPathWithUUID: '/v1/webapi/sites/:clusterId/locks/:uuid',
```
Doesn't our Typescript code generally prefer `locksPathWithUuid` instead?

hatched replied:
Sounds good, updated!

zmb3 commented on web/packages/teleport/src/config.ts:
```diff
@@ -492,6 +497,22 @@ const cfg = {
+    return generatePath(cfg.api.locksPath, { clusterId });
+  },
+
+  getLocksUrlWithUUID(clusterId: string, uuid: string) {
```
Yeah, see how we used `Url` above instead of `URL`? I think the Typescript convention we just is just to capitalize the first letter of an initialism (unlike in Go where we capitalize all of them)

zmb3 commented on web/packages/teleport/src/features.tsx:
```diff
@@ -322,6 +328,44 @@ export class FeatureAuthConnectors implements TeleportFeature {
+  section = ManagementSection.Access;
+
+  route = {
+    title: 'Session & Identity Locks',
```
Let's see what @roraback thinks, but I'm inclined to just say "Locks" here, since "session" is not something you can actually lock.

avatus reviewed (commented):
This feature is really rad! Felt great to use. A couple UX comments. Also, one question. The "quick add user" bar has a white background and, although it sorta fits with the white background of the dropdown on the same screen, should this instead be a dark background input like the rest of the app? NOT A BLOCKER, I'm assuming the designs said white but I'm just curious.

kimlisa commented:
do we have a figma design for this? this lock screen and functions is very similar to access requests screen and am curious why we it's so different. Most of the UI/UX issues i see atm can be resolved by looking at the access request design.

Some other things that might not be obvious in access request:

- providing a confirmation dialog when deleting locks
- providing a alert dialog when user navigates away from the current route, that their current selections will be cleared

feature, possible bug questions:
- was it intentional that we can add non-existing resource with the `quick add resource` input box? i couldn't find this in the docs. i see that `creating` the lock with non-existing resource doesn't create an error (?). but i can see an issue if the user was simply misspelling, so if this wasn't intentional, perhaps we can just provide a client search bar with the table?
- something to think about is that it's possible to actually lock myself (the locker) out by say, locking a role that I am assigned. i wonder if we should maybe do a confirmation dialog for these cases. for resource user and roles is quite easy to do.

hatched commented on web/packages/teleport/src/Locks/NewLock.tsx:
```diff
@@ -0,0 +1,238 @@
+    columns.push({
+      altKey: 'add-btn',
+      render: ({ name }) => (
+        <BtnCell cb={onAdd.bind(null, name)}>+ Add</BtnCell>
```
While there aren't really any negative side effects, no this should be disabled once added to the lock list.

hatched replied:
This button is now disabled and a spinner is shown.

hatched commented on web/packages/teleport/src/Locks/CreateLock.tsx:
```diff
@@ -0,0 +1,171 @@
+        </Box>
+        <Box mt={3}>
+          <Text mr={2}>TTL: </Text>
+          <Input placeholder={`5h`} ref={ttlRef} data-testid="ttl" />
```
I actually ran into this myself because I didn't know all of the suffixes for a TTL. In a follow-up there will be an expansion on this UI where you can input an `expires` date instead of a `ttl`. But I can add some more detail to the placeholder to illustrate all allowed suffixes.

avatus replied:
I think just two examples + "empty = never" would be enough to get the context of how it works. Maybe its already enough given our TTL syntax elsewhere. Just a thought

hatched replied:
Added!

hatched commented on web/packages/teleport/src/Locks/Locks.tsx:
```diff
@@ -0,0 +1,136 @@
+  const { clusterId } = useStickyClusterId();
+  const { locks, fetchLocks } = useLocks(clusterId);
+
+  function onDelete(lockName: string) {
```
done

zmb3 commented:
<img width="733" alt="image" src="https://user-images.githubusercontent.com/7527103/228306842-58bb215a-81f7-4839-9dbd-0f55008cf00f.png">

Looks like the labels aren't rendering correctly

hatched commented:
Fixed the labels issue:

<img width="927" alt="Screenshot 2023-03-28 at 12 11 08 PM" src="https://user-images.githubusercontent.com/532033/228330075-da4a93c5-6d79-4d83-adb2-76869f75dfc1.png">

hatched commented on web/packages/teleport/src/Locks/NewLock.tsx:
```diff
@@ -0,0 +1,238 @@
+}: TargetListProps) {
+  if (!data) data = [];
+
+  if (selectedTarget === 'device') {
```
I left this special handling in but removed it from the list. It'll make it a bit easier to enable later.

avatus commented:
Love all the new loading states and feels really polished and cool. A couple small things.

In the "New Lock" form, we have the field "Message", but in the locks list, the column is titled "Reason". I think I lean more towards "Reason", but whatever we pick it should be consistent across both components.

Love the disabling of the buttons when you add a user. Also, when adding a user from the quickadd, it disables it as well! But adding from the button and then adding that user from the quickadd will allow the user to be added twice. Since this really isn't the expected usecase, as well as nothing really detrimental happens (and you can fix in the New Lock slider anyway) I don't think this is a blocker but if its a quick fix I'd take a look at it.

hatched commented:
> Love all the new loading states and feels really polished and cool. A couple small things.

Thanks!

> In the "New Lock" form, we have the field "Message", but in the locks list, the column is titled "Reason". I think I lean more towards "Reason", but whatever we pick it should be consistent across both components.

I've updated it to be `Message`. I think the design was taking inspiration from access requests which have a "reason" but locks have a "message",

> Love the disabling of the buttons when you add a user. Also, when adding a user from the quickadd, it disables it as well! But adding from the button and then adding that user from the quickadd will allow the user to be added twice. Since this really isn't the expected usecase, as well as nothing really detrimental happens (and you can fix in the New Lock slider anyway) I don't think this is a blocker but if its a quick fix I'd take a look at it.

😁 I'm glad you noticed that nice little touch! Good catch with the inverse though, I indented on that so I've updated it, now if you try and `quick add` the same lock target again it'll disable the `add` button.

avatus reviewed (approved):
Feels great, love the feature

zmb3 reviewed (approved):
Please rebase the commits before adding to merge queue so we don't get a massive autogenerated commit message.
